docs/proposals/hpa/federated-hpa-v2.md

```diff
@@ -0,0 +1,809 @@
+    replicas: 15
+```
+
+So after the assignment, the result of the HPA assignment will be:
+```
+#member 1
+minReplicas: 8
+maxReplicas: 20
+#member 2
+minReplicas: 1
+maxReplicas: 4
+```
+
+And if the sum of availableReplicas is smaller than FederatedHPAMaxReplicas, the left will be assigned to the cluster with the highest priority.
+
+The resource binding will be updated after the assignment, same with `StaticWeighted assignment policy`.
+
+
+### Autoscaling across clusters
+With this design architecture, the behavior of scaling across cluster containers has two parts: Scaling Up/Scaling down.
+And also, to make the FederatedHPA controller scaling work(no conflicting), FederatedHPA controller only can scale up the workload in member clusters.
+
+#### Duplicated assignment policy
+##### Scaling up
+FederatedHPA controller should list&watch the status of member HPA, when one member cluster could not scale up the workload(HPA status will show), the FederatedHPA controller will scale up the workload in other member clusters(the one with the most available replicas).
+
+The number of replicas that need to be scaled up in different clusters will be calculated from the member cluster's HPA status(metrics or pending pods exists)
+
+##### Scaling down
+FederatedHPA controller should not care about this.
+
+#### StaticWeighted assignment policy
+##### Scaling up
+Similar to `Duplicated assignment policy`, but FederatedHPA controller will scale up the replicas in member clusters based on the static weight.
+
+For example, there is the following cluster's status:
+```
+targetUtilization: 50%
+cluster 1 static weight: 1
+cluster 2 static weight: 3
+cluster 3 static weight: 1
+
+cluster 1:
+  current replicas: 4
+  maxReplicas: 4
+  current Utilization: 100
+
+cluster 2:
+  current replicas: 8
+  maxReplicas: 100
+  current Utilization: 50
+
+cluster 3:
+  current replicas: 8
+  maxReplicas: 100
+  current Utilization: 50
+
+```
+
+So, FederatedHPA controller will directly scale 3 replicas up in cluster 2, and 1 replica in cluster 3(If there are pending pods, the behavior will be the same).
+
+##### Scaling down
+FederatedHPA controller should not care about this.
+
+#### DynamicWeighted assignment policy
+##### Scaling up
+Similar to `StaticWeighted assignment policy`, but FederatedHPA controller will scale up the replicas in member clusters based on the dynamic weight(availableReplicas).
+
+For example, there is the following cluster's status:
+```
+targetUtilization: 50%
+
+cluster 1:
+  current replicas: 4
+  maxReplicas: 4
+  current Utilization: 100
+
+cluster 2:
+  current replicas: 8
+  maxReplicas: 100
+  current Utilization: 50
+  currentAvailableReplicas: 50
+
+cluster 3:
+  current replicas: 8
+  maxReplicas: 100
+  current Utilization: 50
+  currentAvailableReplicas: 50
+
+```
+
+So, FederatedHPA controller will directly scale 2 replicas up in cluster 2, and 2 replicas in cluster 3(If there are pending pods, the behavior will be the same).
```

Comment:
Now we need to scale up 4 replicas in member1, but it does not have more available resources. So we assign these 4 replicas to member2 and member3 by weight of their currentAvailableReplicas, right?

How does FederatedHPA controller do this scaling? By calling scale subresource of  workload in member clusters directly? Does it mean FederatedHPA controller should watch all member clusters?

Reply:
> Now we need to scale up 4 replicas in member1, but it does not have more available resources. So we assign these 4 replicas to member2 and member3 by weight of their currentAvailableReplicas, right?

Right.

> By calling scale subresource of workload in member clusters directly?

Yes

> Does it mean FederatedHPA controller should watch all member clusters?

I think FederatedHPA controller could find all the information needed in the work resources.
For updating the scale subresource, we don't need to watch it, just update it directly.

Reply:
> I think FederatedHPA controller could find all the information needed in the work resources.

I guess By fetching the available replicas and spec replicas of this workload? If not equal, we may need to trigger FederatedHPA controller to adjust replicas in member clusters.

> For updating the scale subresource, we don't need to watch it, just update it directly.

It seems like FederatedHPA controller should also have all kubeconfig files because it needs to update the scale subresources.

Reply:
> I guess By fetching the available replicas and spec replicas of this workload?

more specifically, the status of FederatedHPA(contains the utilization of CPU), and the status of the deployment. When one cluster reaches the maxReplicas and utilization is above the target, or the unavailable replicas in deployment status(long time pending), FederatedHPA controller should scale in other clusters

> It seems like FederatedHPA controller should also have all kubeconfig files because it needs to update the scale subresources.

Yes